csp/random.py

```diff
@@ -0,0 +1,167 @@
+        s_last_value = None  # Placeholder for cumulative value between trigger ticks
+
+    if csp.ticked(drift, covariance) and csp.valid(drift, covariance):
+        if s_last_change is None:
+            if not drift.ndim == 1:
+                raise ValueError(f"{csp.now()}: drift must be 1-dimensional")
+            if not covariance.ndim == 2:
+                raise ValueError(f"{csp.now()}: covariance must be 2-dimensional")
+            if covariance.shape[0] != covariance.shape[1]:
+                raise ValueError(f"{csp.now()}: covariance must be a square matrix")
+            if not drift.size == covariance.shape[0]:
+                raise ValueError(f"{csp.now()}: drift and covariance must have same length")
+            s_last_value = np.zeros_like(drift)
+        else:
+            t = (csp.now() - s_last_change).total_seconds()
+            values = _make_brownian_increment(t, s_rng, s_last_drift, s_cov_decomp)
+            np.add(values, s_last_value, out=s_last_value)
+        s_last_change = csp.now()
+        if s_last_drift is not None and drift.shape != s_last_drift.shape:
+            raise ValueError(f"{csp.now()}: shape of drift is not allowed to change")
+        s_last_drift = drift
+
+    if csp.ticked(covariance):
+        # Only do the covariance decomposition when it changes
+        if s_cov_decomp is not None and covariance.shape != s_cov_decomp.shape:
+            raise ValueError(f"{csp.now()}: shape of covariance is not allowed to change")
+        s_cov_decomp = _matrix_decomposition(covariance, csp.now())
+
+    if csp.ticked(trigger) and csp.valid(drift, covariance):
+        if s_last_change is not None:
+            t = (csp.now() - s_last_change).total_seconds()
+            values = _make_brownian_increment(t, s_rng, s_last_drift, s_cov_decomp)
+            # Add in "last value" in case drift/covariance changed in between trigger updates
+            np.add(s_last_value, values, out=values)
+            s_last_change = csp.now()
+            s_last_value.fill(0.0)
+            return values
+
+
+@csp.graph
+def brownian_motion(
+    trigger: ts[object],
+    drift: ts[Numpy1DArray[float]],
+    covariance: ts[NumpyNDArray[float]],
+    seed: object,
+    precise: bool = False,
+) -> ts[NumpyNDArray[float]]:
+    """Generate multi-dimensional Brownian motion at the trigger times, with time-varying drift and covariance.
+
+    Args:
+        trigger: When to return the value of the process
+        drift: Drift parameter (per second), i.e. array of length n
+        covariance: Covariance matrix (per second), array of size nxn
+        seed: The seed for the numpy random Generator. Can be anything accepted by np.random.default_rng
+        precise: Whether to use more precise (i.e. Kahnan) sum in csp.stats.sum when adding increments
+    """
+    increments = brownian_increments(trigger, drift, covariance, seed=seed)
+    return csp.stats.sum(increments, precise=precise)
```

Comment:
I think this is better for instructional purposes on how csp graphs are structured, but if you're going for pure efficiency, it would be faster to just accumulate the Brownian motion directly rather than have a separate increment node. 
1) increment is returning an additional array copy every time its triggered, we could just add it directly to the Brownian motion (s_last_value)
2) we wouldn't have to fill the s_last_value array in increment with 0s every time 
3) the stats node is much more overhead than maintaining the sum as a member

Reply:
I went back and forth on this for a bit, but I think you convinced me. I will add a flag that will return just the increments if desired, as it helps with testing and some other use cases.

Reply:
Discussed offline - let's keep it to one function for Brownian motion and increments with a boolean option. The csp.stats.sum precise option isn't going to make a huge difference